geo/src/algorithm/centroid.rs

```diff
@@ -340,28 +339,136 @@ where
     type Output = Option<Point<T>>;
 
     fn centroid(&self) -> Self::Output {
-        if self.0.is_empty() {
-            return None;
+        centroid_of_coords(self.iter().map(|p| p.0)).map(Point)
+    }
+}
+
+impl<T> Centroid for Geometry<T>
+where
+    T: CoordFloat + FromPrimitive + Sum + Default,
+{
+    type Output = Option<Point<T>>;
+
+    crate::geometry_delegate_impl! {
+        fn centroid(&self) -> Self::Output;
+    }
+}
+
+impl<T> Centroid for GeometryCollection<T>
+where
+    T: CoordFloat + FromPrimitive + Sum + Default,
+{
+    type Output = Option<Point<T>>;
+    fn centroid(&self) -> Self::Output {
+        use crate::algorithm::dimensions::{Dimensions, HasDimensions};
+
+        let mut centroid_0d_accum: Option<(Point<T>, T)> = None;
+        let mut centroid_1d_accum: Option<(Point<T>, T)> = None;
+        let mut centroid_2d_accum: Option<(Point<T>, T)> = None;
+
+        for geometry in self.iter() {
+            let centroid = geometry.centroid();
+            match (geometry.dimensions(), centroid) {
+                (Dimensions::Empty, _) | (_, None) => continue,
+                (Dimensions::ZeroDimensional, Some(centroid)) => {
+                    let mut centroid_accum = centroid_0d_accum.unwrap_or_default();
+                    let weight = T::one();
+                    centroid_accum.0 = centroid_accum.0 + centroid * weight;
+                    centroid_accum.1 = centroid_accum.1 + weight;
+                    centroid_0d_accum = Some(centroid_accum);
+                }
+                (Dimensions::OneDimensional, Some(centroid)) => {
+                    let mut centroid_accum = centroid_1d_accum.unwrap_or_default();
+
+                    fn length<T>(geometry: &Geometry<T>) -> T
+                    where
+                        T: CoordFloat + FromPrimitive + Sum + Default,
+                    {
+                        match geometry {
+                            Geometry::Line(l) => l.euclidean_length(),
+                            Geometry::LineString(l) => l.euclidean_length(),
+                            Geometry::MultiLineString(l) => l.euclidean_length(),
+                            Geometry::GeometryCollection(geometry_collection) => {
+                                geometry_collection.iter().map(length).sum()
+                            }
+                            Geometry::Point(_)
+                            | Geometry::MultiPoint(_)
+                            | Geometry::Polygon(_)
+                            | Geometry::MultiPolygon(_)
+                            | Geometry::Rect(_)
+                            | Geometry::Triangle(_) => T::zero(),
```

Comment:
Do we ever expect this branch to get hit? If not, do you have opinions about using `unreachable!()` here?

Reply:
It's complicated 😅 

For `Geometry::Point(_)`, and `Geometry::MultiPoint(_)`, I'm confident that it should never happen and could add an unreachable.

HasDimensions for Rect and Triangle supports degenerating to 1-D.

```
impl<C: CoordNum> HasDimensions for Triangle<C> {
    ...
    fn dimensions(&self) -> Dimensions {
        if self.0 == self.1 && self.1 == self.2 {
            // degenerate triangle is a point
            Dimensions::ZeroDimensional
        } else if self.0 == self.1 || self.1 == self.2 || self.2 == self.0 {
            // degenerate triangle is a line
            Dimensions::OneDimensional
        } else {
            Dimensions::TwoDimensional
        }
    }
    ...
}
```
So I guess it's expected for some inputs.

For Polygon/MultiPolygon... it's certainly impossible for valid Polygons (closed linestring with 3+ points), but am not confident that we'd want to crash a production instance.

How about a debug_assert for those?

Reply:
> How about a debug_assert for those?

Up to you! My preference is to have assertions (or a comment) since it'll indicate to the reader that this is not a branch we expect to get hit, and could surface bugs